lightkurve/correctors/pldcorrector.py

```diff
@@ -82,199 +116,509 @@ class PLDCorrector(object):
+          components.
+        * Optionally, a single column of ones for numerical stability.
+
+        Thus, the shape of the design matrix will be
+        (n_cadences, n_pld_mask_pixels + n_pca_terms*(pld_order-1) + include_column_of_ones)
 
-    def correct(self, aperture_mask=None, cadence_mask=None, gp_timescale=30,
-                use_gp=True, pld_order=2, n_pca_terms=10, pld_aperture_mask=None):
-        r"""Returns a PLD systematics-corrected LightCurve.
+        Returns
+        -------
+        design_matrix : 2D numpy array
+            See description above.
+        """
+        # We use an optional dependency for very fast PCA (fbpca), but if the
+        # import fails we will fall back on using the slower `np.linalg.svd`.
+        use_fbpca = True
+        try:
+            from fbpca import pca
+        except ImportError:
+            use_fbpca = False
+            log.warning("PLD systematics correction will run faster if the "
+                        "optional `fbpca` package is installed "
+                        "(`pip install fbpca`).")
+
+        matrix_sections = []  # list to hold the design matrix components
+        first_order_matrix = self.create_first_order_matrix()
+
+        # Input validation: n_pca_terms cannot be larger than the number of cadences
+        n_cadences = len(first_order_matrix)
+        if n_pca_terms > n_cadences:
+            log.warning("`n_pca_terms` ({}) cannot be larger than the number of cadences ({});"
+                        "using n_pca_terms={}".format(n_pca_terms, n_cadences, n_cadences))
+            n_pca_terms = n_cadences
+
+        # The original EVEREST paper includes a column vector of ones in the
+        # design matrix to improve the numerical stability (see Luger et al.);
+        # it is unclear whether this is necessary, so this is an optional step
+        # for now.
+        if include_column_of_ones:
+            matrix_sections.append([np.ones((len(first_order_matrix), 1))])
+
+        # Add the first order matrix
+        matrix_sections.append(first_order_matrix)
+
+        # Add the higher order PLD design matrix columns
+        for order in range(2, pld_order + 1):
+            # Take the product of all combinations of pixels; order=2 will
+            # multiply all pairs of pixels, order=3 will multiple triples, etc.
+            matrix = np.product(list(multichoose(first_order_matrix.T, order)), axis=1).T
+            # This product matrix becomes very big very quickly, so we reduce
+            # its dimensionality using PCA.
+            if use_fbpca:  # fast mode
+                components, _, _ = pca(matrix, n_pca_terms)
+            else:  # slow mode
+                components, _, _ = np.linalg.svd(matrix)
+            section = components[:, :n_pca_terms]
+            matrix_sections.append(section)
+
+        return np.concatenate(matrix_sections, axis=1)
+
+    def create_pymc_model(self, design_matrix=None, cadence_mask=None,
```

Comment:
I can pass in an arbitrary design matrix here. This is super. If I understand correctly, in theory then I can pass in e.g. the CBVs, and get out the best correction.

Reply:
👍👍👍